### ISSUE TITLE
Define instruction outputs for non-derivable capability inputs

### ISSUE THREAD

**Issue by andresag01** [CTSRD-CHERI/cheri-specification#68]
CHERI instructions, like `CGetBase`, `CSetBounds`, etc, take a capability as an input to produce some output. It is possible that the input is some arbitrary value that does not encode a valid capability. What is the expected output in this case? Ideally, the output is well-defined to facilitate activities like verification.

For example, a program in RV64 could write an arbitrary 16-byte sequence to memory and load that "malformed" capability using `LC`. The tag of that capability would of course be cleared, but it can be provided as an input to  (e.g.) `CSetBounds` or `CGetBase`. What should be the output?

By malformed, I mean that the capability is not _derivable_. In other words, that capability could not have been legally constructed if we started with the almighty capability and used only legal instructions like `CSetBounds`, `CSetAddr`, `CSetPerms`, etc. 

There are proposals to deal with this problem:

1. The input capability is _legalized_ meaning that the capability is somehow re-interpreted to fix a broken encoding. For example, forcing IE=0 when the exponent is greater than 52 as its done [here](https://github.com/CTSRD-CHERI/sail-cheri-riscv/blob/f3a264c1966fcdc3005f889abdd40ff293b89ab8/src/cheri_cap_common.sail#L248).
1. The output of `CGetBase`, `CGetLen`, etc is an error value, like 0, when the input capability is not derivable.

Note that either proposal relies on having a clear definition of what it means to be derivable. For example, is a capability with the non-zero reserved bits derivable?

**Comment by PeterRugg** [comment]
Hmm, it seems like your check still doesn't line up with your definition of "derivable": there are lots of capabilities with e <= 51 whose top goes past the end of the address space, e.g. a 2-byte capability with its base pointing to 2^64-1. That said, I don't think we need to legalise/error out on those cases, but then why check the top at all?

**Comment by tariqkurd-repo** [comment]
So there are more cases there the top exceeds the address space - I guess we should detect them all?

**Comment by andresag01** [comment]
The problem is the definition of "derivable". You can still use the regular setCapAddress, setCapBounds, etc operations to derive the capability that Peter described even thought that capability's top is actually 0x1_0000_0000_0000_0001, which still fits because the top is 65 bits although you cannot access the memory location at 0x1_0000_0000_0000_0000 that's technically "in bounds". But should you include things like CSetHigh in the definition of derivable, etc? Maybe not? Still needs thought!

**Comment by jrtc27** [comment]
CSetHigh clears the tag; it’s no different to just writing whatever bits you want to memory other than not requiring the round-trip through memory.

**Comment by tariqkurd-repo** [comment]
Yes - what I think we're trying to detect is cases where the bounds are mathematically unexpected, as oppose to just not very useful. As @jrtc27 says, CSetHigh can do anything, but things like CGetBase need to return something which can be usefully interpreted by the software, and not _junk_ due to an out of range shift truncating upper bits - which is what I think we're trying to detect.

Given that - are there any perceived problems with the definition above that @andresag01 suggests?

**Comment by PeterRugg** [comment]
Hmm, the check I had proposed was similar, but without any condition on the top. The proposal here is too strict for "all possibly tagged things", as discussed. I think it's also stricter than the "all things that are reachable using CSetBounds,... without necessarily being tagged": surely you can CSetBounds a capability with address -1 asking for length -1 and it will round and set bit 65. My proposal was just to ignore the top, legalise/error out for capabilities where the E>52 or base bits overflowed past 2^64, and make the arithmetic work for all other capabilities. I think I needed 66-bit arithmetic, which this proposal might avoid. However, checking the top means the error case can't be detected until you've rederived the top bits of T, which means doing a mantissa-width subtract, whereas the conditions I proposed were much quicker, though I'm not sure if this will ever be on a critical path. A simpler condition might also make it easier to reuse this wasted space in future?

**Comment by andresag01** [comment]
I do not think checking the expanded top is in the critical path. The check could be performed fairly late leaving sufficient time to compute top.

**Comment by andresag01** [comment]
As discussed, the instructions (e.g. CGetBase, CGetLen, etc) perform the following check:
```
allowed = !IE || (E == 52 && B[13:12] == 0) || (E == 51 && B[13] == 0) || (E <= 50)
```
There are to implementation options considered so far when `!allowed`:
1. Assume that `IE == 0` and output whatever the instruction would normally do
2. Output `0` or some other constant value to indicate "failure"

**Comment by tariqkurd-repo** [comment]
Similar to the way that `CGetPerm` isolates the software from the exact hardware format for storing permissions, I think that a `CIsDerivable` instruction would be useful for interpreting results from CGetBase etc. otherwise the test on E and B[13:12] etc. might be run an awful lot, and any changes to the derivability check would directly affect the software.

**Comment by PeterRugg** [comment]
Apologies, went on holiday just after this. It would certainly be an easy instruction to provide... I wonder why anybody would ever use it though. Presumably it would live deep down in a few error diagnosing functions?

**Comment by tariqkurd-repo** [comment]
yes - it's a debug thing - the point is to abstract the capability format into the hardware implementation - otherwise the exact details need to be know by software. Maybe that's the case anyway (e.g. the compiler needs to know how to encode caps), but the application software or kernel probably don't need to know the details at all, until an non-derivable cap appears and it needs to work out what went wrong. Are there other reasons why software would need to understand the format? Why do we have `CGetPerm` which adds a hardware abtraction layer to the permission format?

**Comment by PeterRugg** [comment]
I agree that abstracting like this is good. You could see the "derivability" as being abstracted behind CBuildCap, which refuses to tag underivable caps, and CGetBase + CGetLen, which treat them in some defined way. I guess I'm saying beyond abstracting away the definition of which capabilities are derivable, we could abstract away the fact that underivable capabilities exist at all...
But yes, if it's a case that we think debug will want to check for, it's probably worth having the instruction or considering another string of operations that could emulate it. This could tie in well with https://github.com/CTSRD-CHERI/cheri-specification/issues/117: if we do what you're suggesting there then zero-length capabilities become "underivable", and we can just use `CGetLen == 0` to mean derivable.

**Comment by tariqkurd-repo** [comment]
I think that's an excellent idea

**Comment by jrtc27** [comment]
Software outside of a debugger, which already needs to and does know the format, doesn’t care about derivability, it cares about whether it can be derived from a specific root. Which is called “does CBuildCap work?” I struggle to see a use case for CIsDerivable.
